internal/providers/cloudstack/cloudstack.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The CloudStack provider fetches configurations from the userdata available in
+// the config-drive.
+// NOTE: This provider is still EXPERIMENTAL.
+
+package cloudstack
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/coreos/ignition/config"
+	"github.com/coreos/ignition/config/types"
+	"github.com/coreos/ignition/config/validate/report"
+	"github.com/coreos/ignition/internal/log"
+	"github.com/coreos/ignition/internal/resource"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	diskByLabelPath         = "/dev/disk/by-label/"
+	configDriveUserdataPath = "/cloudstack/userdata/user_data.txt"
+)
+
+func FetchConfig(logger *log.Logger, client *resource.HttpClient) (types.Config, report.Report, error) {
+	var data []byte
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+
+	dispatch := func(name string, fn func() ([]byte, error)) {
+		raw, err := fn()
+		if err != nil {
+			switch err {
+			case context.Canceled:
+			case context.DeadlineExceeded:
+				logger.Err("timed out while fetching config from %s", name)
+			default:
+				logger.Err("failed to fetch config from %s: %v", name, err)
+			}
+			return
+		}
+
+		data = raw
+		cancel()
+	}
+
+	go dispatch("config drive (config)", func() ([]byte, error) {
+		return fetchConfigFromDevice(logger, ctx, "config")
+	})
+
+	go dispatch("config drive (CONFIG)", func() ([]byte, error) {
+		return fetchConfigFromDevice(logger, ctx, "CONFIG")
+	})
+
+	<-ctx.Done()
+	if ctx.Err() == context.DeadlineExceeded {
+		logger.Info("Config drive was not available in time. Continuing without a config...")
+	}
+
+	return config.Parse(data)
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return (err == nil)
+}
+
+func labelExists(label string) bool {
+	_, err := getPath(label)
+	return (err == nil)
+}
+
+func getPath(label string) (string, error) {
+	path := diskByLabelPath + label
+
+	if fileExists(path) {
+		return path, nil
+	}
+
+	out, err := exec.Command("blkid", "-l", "-t", "LABEL="+label, "-o", "device").Output()
```

Comment:
We added it as a backup. But normally it should not be needed indeed, so we removed it.